File: Src/Engine/Editor/Modules/UIModule.cpp

#include "UIModule.h"

#include "WindowsModule.h"
#include "Editor/EditorApp.h"
#include "Editor/GUI/MainMenu.h"
#include "editor/gui/mainmenubutton.h"
#include "Editor/GUI/StatusBar.h"
#include "Editor/GUI/ToolStrip.h"
#include "Editor/GUI/ContextMenu/ContextMenu.h"
#include "Editor/GUI/Docking/MasterDockPanel.h"
#include "Editor/Resource/Thumbnails/ThumbnailRequest.h"
#include "Editor/Resource/Thumbnails/Thumbnails.h"
#include "Editor/Windows/ContentWindow.h"
#include "Runtime/UI/GUI/Style.h"
#include "Runtime/UI/GUI/WindowRootControl.h"

#include "Editor/Test.h"

namespace SE::Editor
{
	int UIModule::InitOrder()
	{
		return -95;
	}

	UIModule::UIModule(EditorApp* editor) : EditorModule(editor)
	{
		thumbnails = New<Thumbnails>(editor);
	}

	void UIModule::OnInit()
	{
		m_MasterPanel = New<MasterDockPanel>();
		auto mainWindow = editor->GetMainWindow()->GetGUI();

		// Update window background
		mainWindow->BackgroundColor = Style::Current->Background;

		thumbnails->OnInit();

		InitMainMenu(mainWindow);
		InitToolstrip(mainWindow);
		InitStatusBar(mainWindow);
		InitDockPanel(mainWindow);
	}

	void UIModule::OnUpdate()
	{
		thumbnails->OnUpdate();
	}
	void UIModule::OnExit()
	{
		thumbnails->OnExit();
	}

	MasterDockPanel* UIModule::GetMasterPanel()
	{
		return m_MasterPanel;
	}

	void UIModule::InitMainMenu(RootControl* mainWindow)
	{
		MainMenu = New<Editor::MainMenu>(mainWindow);
		MainMenu->Parent = mainWindow;

		ContextMenu* contextMenu = nullptr;

		// Window
		MainMenuButton* menuWindow = MainMenu->AddButton(SE_TEXT("Window"));
		contextMenu = menuWindow->ContextMenu;
		// contextMenu->VisibleChanged += OnMenuWindowVisibleChanged;
		contextMenu->AddButton(SE_TEXT("Content"), CreateFunc([this]()
		{
			editor->windowsModule->ContentWin->FocusOrShow();
		}));
		contextMenu->AddButton(SE_TEXT("TestScene"), CreateFunc([this]()
		{
			CreateTestScene();
		}));
		/*contextMenu->AddButton(SE_TEXT("Scene"), Editor.Windows.SceneWin.FocusOrShow);
		contextMenu->AddButton(SE_TEXT("Toolbox"), Editor.Windows.ToolboxWin.FocusOrShow);
		contextMenu->AddButton(SE_TEXT("Properties"), Editor.Windows.PropertiesWin.FocusOrShow);
		contextMenu->AddButton(SE_TEXT("Game"), Editor.Windows.GameWin.FocusOrShow);
		contextMenu->AddButton(SE_TEXT("Editor"), Editor.Windows.EditWin.FocusOrShow);
		contextMenu->AddButton(SE_TEXT("Debug Log"), Editor.Windows.DebugLogWin.FocusOrShow);
		contextMenu->AddButton(SE_TEXT("Output Log"), Editor.Windows.OutputLogWin.FocusOrShow);
		contextMenu->AddButton(SE_TEXT("Graphics Quality"), Editor.Windows.GraphicsQualityWin.FocusOrShow);
		contextMenu->AddButton(SE_TEXT("Game Cooker"), Editor.Windows.GameCookerWin.FocusOrShow);
		contextMenu->AddButton(SE_TEXT("Profiler"), inputOptions.ProfilerWindow, Editor.Windows.ProfilerWin.FocusOrShow);
		contextMenu->AddButton(SE_TEXT("Content Search"), Editor.ContentFinding.ShowSearch);
		contextMenu->AddButton(SE_TEXT("Visual Script Debugger"), Editor.Windows.VisualScriptDebuggerWin.FocusOrShow);*/
		contextMenu->AddSeparator();
		// contextMenu->AddButton("Save window layout", Editor.Windows.SaveLayout);
		/*_menuWindowApplyWindowLayout = contextMenu->AddChildMenu("Window layouts");
		cm.AddButton("Restore default layout", Editor.Windows.LoadDefaultLayout);*/
	}

	void UIModule::InitToolstrip(RootControl* mainWindow)
	{
		m_ToolStrip = New<ToolStrip>(34.0f, MainMenu->Bottom);
		m_ToolStrip->Parent = mainWindow;
	}

	void UIModule::InitStatusBar(RootControl* mainWindow)
	{
		// Status Bar
		m_StatusBar = New<StatusBar>();
		m_StatusBar->Text = "Loading...";
		m_StatusBar->Parent = mainWindow;
		m_StatusBar->Offsets = Margin(0, 0, -StatusBar::DefaultHeight, StatusBar::DefaultHeight);
	}

	void UIModule::InitDockPanel(RootControl* mainWindow)
	{
		// Dock Panel
		m_MasterPanel->AnchorPreset = AnchorPresets::StretchAll;
		m_MasterPanel->Parent = mainWindow;
		m_MasterPanel->Offsets = Margin(0, 0, m_ToolStrip->Bottom, m_StatusBar->Height);
	}
} // SE

File: Src/Engine/Editor/Modules/UIModule.cs
using System;
using SE.Editor.GUI;
using SE.GUI;

namespace SE.Editor
{
    /// <summary>
    /// Owns the managed editor chrome and the master docking workspace.
    /// </summary>
    public sealed class UIModule : EditorModule
    {
        private WindowRootControl? m_Root;
        private float m_LastWidth = float.NaN;
        private float m_LastHeight = float.NaN;

        public UIModule(Editor editor) : base(editor)
        {
        }

        public MainMenu MainMenu { get; private set; } = null!;
        public ToolStrip ToolStrip { get; private set; } = null!;
        public StatusBar StatusBar { get; private set; } = null!;
        public MasterDockPanel MasterDockPanel { get; private set; } = null!;

        internal override int Order => -95;

        public override void OnInit()
        {
            m_Root = Editor.MainWindow.GUI;
            MainMenu = m_Root.AddChild(new MainMenu(Math.Max(m_Root.Width, 1.0f)));
            ToolStrip = m_Root.AddChild(new ToolStrip(ToolStrip.DefaultMarginV * 2.0f + 26.0f, MainMenu.DefaultHeight, Math.Max(m_Root.Width, 1.0f)));
            StatusBar = m_Root.AddChild(new StatusBar(0.0f, Math.Max(m_Root.Width, 1.0f)));
            MasterDockPanel = m_Root.AddChild(new MasterDockPanel());

            BuildMenus();
            BuildToolStrip();
            StatusBar.Text = "Ready";
            UpdateLayout(force: true);
        }

        public override void OnUpdate()
        {
            UpdateLayout(force: false);
            ThumbnailService.Instance.Update();
        }

        public override void OnDispose()
        {
            if (m_Root == null)
                return;

            DisposeControl(MasterDockPanel);
            DisposeControl(StatusBar);
            DisposeControl(ToolStrip);
            DisposeControl(MainMenu);
            m_Root = null;
        }

        private void BuildMenus()
        {
            MainMenuButton window = MainMenu.AddButton("Window");
            window.ContextMenu.AddButton("Content", () => Editor.Windows.Content.FocusOrShow());
            window.ContextMenu.AddButton("Scene Hierarchy", () => Editor.Windows.SceneHierarchy.FocusOrShow());
            window.ContextMenu.AddButton("Log", () => Editor.Windows.Log.FocusOrShow());
            window.ContextMenu.AddButton("Edit Scene", () => Editor.Windows.EditScene.FocusOrShow());
        }

        private void BuildToolStrip()
        {
            ToolStrip.AddButton("Content", () => Editor.Windows.Content.FocusOrShow());
            ToolStrip.AddButton("Scene", () => Editor.Windows.SceneHierarchy.FocusOrShow());
            ToolStrip.AddButton("Log", () => Editor.Windows.Log.FocusOrShow());
        }

        private void UpdateLayout(bool force)
        {
            if (m_Root == null || m_Root.IsDisposed)
                return;

            float width = Math.Max(m_Root.Width, 1.0f);
            float height = Math.Max(m_Root.Height, MainMenu.DefaultHeight + ToolStrip.Height + StatusBar.DefaultHeight + 1.0f);
            if (!force && width == m_LastWidth && height == m_LastHeight)
                return;

            m_LastWidth = width;
            m_LastHeight = height;
            MainMenu.SetBounds(0.0f, 0.0f, width, MainMenu.DefaultHeight);
            float toolStripTop = MainMenu.Y + MainMenu.Height;
            ToolStrip.SetBounds(0.0f, toolStripTop, width, ToolStrip.Height);
            StatusBar.SetBounds(0.0f, height - StatusBar.DefaultHeight, width, StatusBar.DefaultHeight);
            float workspaceTop = ToolStrip.Y + ToolStrip.Height;
            MasterDockPanel.SetBounds(0.0f, workspaceTop, width, Math.Max(1.0f, StatusBar.Y - workspaceTop));
            m_Root.PerformLayout(true);
        }

        private static void DisposeControl(Control? control)
        {
            if (control == null || control.IsDisposed)
                return;

            control.Parent?.RemoveChild(control);
            control.Dispose();
        }
    }
}
